blog/2024-05-08-prefixes.mdx

```diff
@@ -0,0 +1,132 @@
+we unearth pivotal moments that introduced and refined the various prefixes
+integral to `conda`'s architecture. The history is also revealing in how
+non-linear the development of the prefix system has been, explaining many of the
+complexities of trying to wrestle with the prefixes or trying to find ways to
+simplify them.
+
+1. `conda.base.context.context.conda_prefix`
+
+   Introduced in `conda 4.2.0` (see https://github.com/conda/conda/commit/6e730cc)
+   as part of the private environment effort which has since been abandoned and
+   deprecated.
+
+   The `conda_prefix` lies at the heart of the prefix system, denoting the
+   installation directory of `conda` itself. Typically, this aligns with the
+   location of the `base` environment and is equivalent to Python's `sys.prefix`
+   value.
+
+   :::info
+
+   Use this anytime you need to interact with the `conda` installation itself.
+
+   :::
+
+2. `conda.base.context.context.root_prefix`
+
+   > Previously `conda.config.root_dir` → `conda.config.ROOT_DIR`.
+
+   Introduced in `conda 1.1.0` (see https://github.com/conda/conda/commit/d2b370a)
+   and was originally intended for testing purposes only.
+
+   The `root_prefix` serves as a [configurable value][condarc], most easily
+   configurable in the user's `~/.condarc` file. In the absence of user-defined
+   preferences, it defaults to the `conda_prefix`.
+
+   :::warning
+
+   Avoid using this. This is a contender for deprecation.
+
+   :::
+
+3. `conda.base.context.context.active_prefix`
+
+   Introduced in `conda 4.4.0` (see https://github.com/conda/conda/commit/cbbe81f)
+   as part of cleaning up the `conda info` output.
+
+   Reflecting the currently active environment, the `active_prefix` is determined
+   by the `$CONDA_PREFIX` environment variable which is set by the `conda activate`
+   and `conda deactivate` commands.
+
+   :::warning
+
+   Avoid using this. Treat this as an internal value.
+
+   :::
+
+4. `conda.base.context.context.default_prefix`
+
+   > Previously `conda.config.default_prefix` → `conda.config.DEFAULT_ENV_PREFIX` →
+   > `conda.config.default_environment` → `conda.anaconda.default_environment`.
+
+   Introduced in `conda 1.1.0` (see https://github.com/conda/conda/commit/acd8144)
+   as part of a larger configuration overhaul.
+
+   The `default_prefix` effectively aggregates `root_prefix` and `active_prefix`,
+   embodying the default environment in the absence of CLI overrides.
+
+   :::warning
+
+   Avoid using this. Treat this as an internal value.
+
+   :::
+
+5. `conda.base.context.context.target_prefix`
+
+   > Previously `conda.base.context.context.prefix` →
+   > `conda.cli.common.get_prefix()` → `conda.cli.utils.get_prefix()`.
+
+   Introduced in `conda 1.3.0` (see https://github.com/conda/conda/commit/9fd0f4b)
+   to abstract out the `--name`/`--prefix` logic from the CLI implementations.
+
+   The `target_prefix` is the final prefix and first reflects the user input via
```

Comment:
The final prefix in your list of prefixes? In terms of versioning, it seems like it was introduced third.

I also feel like I'm reading final like "ultimate" here.

Reply:
Ah so the prefixes are discussed in order of importance, from the most basic/bottom of the stack to the highest level/top of the stack. So "final" here just indicates that this is the top of the stack. Yes, this is the "ultimate" prefix but that also seems too grand?

Reply:
Oh, definitely, don't use "ultimate" here. I was just saying that's what the connotation felt like to me. And I didn't realize we were discussing them from bottom of the stack to top! That's cool. Maybe I missed that in your explanation?